src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+// impSpanEqualsOrStartsWith: The main entry-point for [ReadOnly]Span<char> methods
+//    We're going to unroll & vectorize the following cases:
+//    1) MemoryExtensions.SequenceEqual<char>(var, "cns")
+//    2) MemoryExtensions.SequenceEqual<char>("cns", var)
+//    3) MemoryExtensions.Equals(var, "cns", StringComparison.Ordinal)
+//    4) MemoryExtensions.Equals("cns", var, StringComparison.Ordinal)
+//    5) MemoryExtensions.StartsWith<char>("cns", var)
+//    6) MemoryExtensions.StartsWith<char>(var, "cns")
+//    7) MemoryExtensions.StartsWith("cns", var, StringComparison.Ordinal)
+//    8) MemoryExtensions.StartsWith(var, "cns", StringComparison.Ordinal)
+//
+// Arguments:
+//    startsWith  - Is it StartsWith or Equals?
+//    sig         - signature of StartsWith or Equals method
+//    methodFlags - its flags
+//
+// Returns:
+//    GenTree representing vectorized comparison or nullptr
+//
+GenTree* Compiler::impSpanEqualsOrStartsWith(bool startsWith, CORINFO_SIG_INFO* sig, unsigned methodFlags)
+{
+    const bool isStatic  = methodFlags & CORINFO_FLG_STATIC;
+    const int  argsCount = sig->numArgs + (isStatic ? 0 : 1);
+
+    GenTree* op1;
+    GenTree* op2;
+    if (argsCount == 3) // overload with StringComparison
+    {
+        if (!impStackTop(0).val->IsIntegralConst(4)) // StringComparison.Ordinal
+        {
+            // TODO: Unroll & vectorize OrdinalIgnoreCase
+            return nullptr;
+        }
+        op1 = impStackTop(2).val;
+        op2 = impStackTop(1).val;
+    }
+    else
+    {
+        op1 = impStackTop(1).val;
+        op2 = impStackTop(0).val;
+    }
+
+    // For generic StartsWith and Equals we need to make sure T is char
+    if (sig->sigInst.methInstCount != 0)
+    {
+        assert(sig->sigInst.methInstCount == 1);
+        CORINFO_CLASS_HANDLE targetElemHnd = sig->sigInst.methInst[0];
+        CorInfoType          typ           = info.compCompHnd->getTypeForPrimitiveValueClass(targetElemHnd);
+        if ((typ != CORINFO_TYPE_SHORT) && (typ != CORINFO_TYPE_USHORT) && (typ != CORINFO_TYPE_CHAR))
+        {
+            return nullptr;
+        }
+    }
+
+    // Try to obtain original string literals out of span arguments
+    GenTreeStrCon* op1Str = impGetStrConFromSpan(op1);
+    GenTreeStrCon* op2Str = impGetStrConFromSpan(op2);
+
+    if (!((op1Str != nullptr) ^ (op2Str != nullptr)))
+    {
+        // either op1 or op2 has to be '(ReadOnlySpan)"cns"'
+        return nullptr;
+    }
+
+    GenTree*       spanObj;
+    GenTreeStrCon* cnsStr;
+    if (op1Str != nullptr)
+    {
+        cnsStr  = op1Str;
+        spanObj = op2;
+    }
+    else
+    {
+        cnsStr  = op2Str;
+        spanObj = op1;
+    }
+
+    int             cnsLength = -1;
+    const char16_t* str       = nullptr;
+    if (cnsStr->IsStringEmptyField())
+    {
+        // check for fake "" first
+        cnsLength = 0;
+        JITDUMP("Trying to unroll String.Equals(op1, \"\")...\n", str)
```

Comment:
Should this be changed to MemoryExtensions?

Reply:
@MichalPetryka good point! thanks, fixed